### ISSUE TITLE
transcripts: project/branch should be validated?

### ISSUE THREAD

**Issue by ceedubs** [unisonweb/unison#5173]
**Is your feature request related to a problem? Please describe.**

I think that the current transcript behavior in `ucm` blocks is that a line like `project/branch> view foo` causes ucm to implicitly use the `project/branch` namespace without an explicit `switch`. This probably makes sense for the initial `ucm` command in a transcript, as you need to start _somewhere_.

However, I think that the implicit switching and project creation makes it awkward to use transcripts to test certain functionality. For example, imagine that you want to test that `project.rename` keeps you in the same branch after the rename. You might write something like this:

```
proj/my-feature> project.rename proj2
proj2/my-feature> ls
```

But this won't really assert that `project.rename` kept you in the `/my-feature` branch, because leading the line with `proj2/my-feature` will implicitly switch you to the `my-feature` branch even if an interactive ucm session wouldn't.

**Describe the solution you'd like**

I think that it's fine for the initial `ucm` command in a transcript to implicitly declare a project. But for all subsequent commands I think that it might make more sense to make the project/branch specifier an assertion to validate, and that switching projects/branches should require an explicit `switch` in a transcript just as it does in ucm.

The more that the transcript and interactive session behavior diverge, the less helpful transcripts are in detecting bugs.

**Describe alternatives you've considered**

I haven't thought a lot about alternatives. But I think that there should be a pretty compelling reason for each divergence of behavior between transcripts and actual ucm sessions.

**Additional context**

This also came up [here](https://github.com/unisonweb/unison/pull/5100#issuecomment-2175832159) when there wasn't a way to add a regression test for remembering the active branch in a project.

**Comment by mitchellwrosen** [comment]
For historical context, we've always had a transcript parser that let you do things like

```
.> whatver
.a> whatever
```

without an explicit `cd`. When we added projects, we also added implicit `switch` behavior matching that.

And then very recently, we even allowed you to create a new branch just by mentioning its name in a prompt! So, the implicit magic / diverging from a real UCM session is worse than you describe in the ticket. And it's true that certain things are impossible to test in transcripts with the auto-switch and auto-branch features. That's not great.

I could personally go either way here, FWIW. Transcripts are quicker than ever to write now, which matters a little.

**Comment by aryairani** [comment]
Related: I was thinking recently that `scratch/main>` is a lot to type to start each and every line in the transcript.

So how about something like:

1. `>` defaults to the "current branch", which is `scratch/main` initially, but can be altered by other commands. the actual branch is shown in the output.md
2. `/branch` or `proj/branch>` implicitly creates-empty and switches to `/branch` or `proj/branch` respectively; this rule can be kept or we can get rid of it
3. `>` or `/branch` or `project/branch>` alone on a line is valid; the current branch or specified branch is shown in the output.md as well

e.g.

---
Input:
````
```ucm
> ls
> branch.create foo
>
/bar>
other/other>
```
````

Output:
````
```ucm
scratch/main> ls
  
  nothing to show

scratch/main> branch.create foo

  Done. I've created the foo branch based off of main.
  
  Tip: To merge your work back into the main branch, first `switch /main` then `merge /foo`.

scratch/foo>
scratch/bar>
other/other>
```
````

4. Currently we suppress the normal implicit branch creation output, but we could decide to include it in the output

**Comment by ceedubs** [comment]
@aryairani I think that my honest answer is that saving a bit of time copy/pasting the prompt from the line above doesn't seem worth it to me. I think that it's handy to have transcripts that show user exactly what a session would look like. And also the more special-cased logic that they have, the more chance that they don't emulate actual ucm behavior (and therefore don't catch a bug).

I do write a decent number of transcripts for bug reports, but admittedly I don't write as many as people who work on ucm. So I might just not feel the pain of writing prompts in transcripts enough. I'm not _strongly_ opposed to the proposal; I just question whether the additional complexity holds its weight.

**Comment by aryairani** [comment]
@ceedubs To clarify, the output is still supposed to show what the session would look like; the input never has. Not sure if this changes your answer.

**Comment by mitchellwrosen** [comment]
> Related: I was thinking recently that `scratch/main>` is a lot to type to start each and every line in the transcript.

I don't think this is sufficient justification for adding more features to the transcript parser – the less code the better (of course) – and a little copy+pasting of prompts while writing transcripts isn't so bad :)

**Comment by aryairani** [comment]
@ceedubs The way we use transcripts currently is the the input file specifies some user processes, and the output file specifies ucm's behavior under those instructions. If there's  need to confirm ucm's behavior, or share it as a tutorial, we'd use the output file for that.

This feature request sounds like a change to make the input file specify ucm's output, and the transcript runner verifies that the output matches the input, but I find that contrary to the intended design and don't think we should do that.

I do agree that it would be good to be able to write demo transcripts and regression transcripts for all ucm behaviors, though.

Was there additional motivation for this feature request besides imagining being able to test `project.rename` or `switch`; e.g. a concrete problem you encountered?

**Comment by ceedubs** [comment]
@aryairani yes there is one linked in the additional context section of the report :)

**Comment by ceedubs** [comment]
@aryairani well I guess that it was concrete in the sense that we weren't able to test that `switch` properly returns to the previous branch in Mitchell's PR. But I'm not sure what you mean by concrete.

**Comment by aryairani** [comment]
@ceedubs Yeah, by concrete, I meant something like that. " I wanted to write a transcript to test `switch`, but couldn't." The `switch` issue was listed under additional context as "this also came up", and the `project.rename` was described as "imagine you want" instead of "I want", so I didn't know if there was another primary non-hypothetical thing you had tried to do but couldn't.

But regardless, I agree it would be good to be able to test such things.

**Comment by sellout** [comment]
You can kind of test this sort of thing now, since #5826.

```` markdown
``` unison :hide
OnMyFeatureBranch = ()
```

``` ucm
proj/my-feature> add

  Okay, I'm searching the branch for code that needs to be
  updated...

  Done.

> projects

  1. proj
  2. scratch

> find OnMyFeatureBranch

  1. OnMyFeatureBranch : ()

> project.rename proj2

  Ok, I renamed proj to proj2.

> projects

  1. proj2
  2. scratch
```

Test that `project.rename` put us in the new project, by checking it has the definition we added.

``` ucm
> find OnMyFeatureBranch

  1. OnMyFeatureBranch : ()
```

And make sure that the definition’s in the context we expect:

``` ucm
proj2/my-feature> find OnMyFeatureBranch

  1. OnMyFeatureBranch : ()
```

And that if we use the original project name, it creates a fresh one without our definition:

``` ucm :error
proj/my-feature> find OnMyFeatureBranch

  ☝️

  I couldn't find matches in this namespace, searching in
  'lib'...

  😶

  No results. Check your spelling, or try using tab completion
  to supply command arguments.

  `debug.find.global` can be used to search outside the current
  namespace.
```

And just to make sure that this definition doesn’t exist in the other project:

``` ucm :error
scrach/main> find OnMyFeatureBranch

  ☝️

  I couldn't find matches in this namespace, searching in
  'lib'...

  😶

  No results. Check your spelling, or try using tab completion
  to supply command arguments.

  `debug.find.global` can be used to search outside the current
  namespace.
```
````

**Comment by ceedubs** [comment]
Thanks @sellout that seems good enough to me and it seems like the approach that I suggested isn't palatable to others, so I'm going to go ahead and close this issue out.
